packages/node/warpsync/syncing.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/iotaledger/goshimmer/packages/core/tangle"
+	"github.com/iotaledger/goshimmer/packages/node/database"
+	"github.com/iotaledger/goshimmer/packages/node/p2p"
+	wp "github.com/iotaledger/goshimmer/packages/node/warpsync/warpsyncproto"
+	"github.com/iotaledger/goshimmer/plugins/epochstorage"
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/generics/lo"
+	"github.com/iotaledger/hive.go/types"
+	"golang.org/x/crypto/blake2b"
+)
+
+type epochSyncStart struct {
+	ei          epoch.Index
+	ec          epoch.EC
+	blocksCount int64
+}
+
+type epochSyncBatchBlock struct {
+	ei    epoch.Index
+	ec    epoch.EC
+	block *tangle.Block
+	peer  *peer.Peer
+}
+
+type epochSyncEnd struct {
+	ei                epoch.Index
+	ec                epoch.EC
+	stateMutationRoot epoch.MerkleRoot
+	stateRoot         epoch.MerkleRoot
+	manaRoot          epoch.MerkleRoot
+}
+
+type blockReceived struct {
+	block *tangle.Block
+	peer  *peer.Peer
+}
+
+func (m *Manager) SyncRange(ctx context.Context, startEpoch, endEpoch epoch.Index, startEC epoch.EC, ecChain map[epoch.Index]*epoch.ECRecord) error {
+	if m.syncingInProgress {
+		return fmt.Errorf("epoch syncing already in progress")
+	}
+
+	if m.isStopped {
+		return fmt.Errorf("warpsync manager is stopped")
+	}
+
+	m.syncingInProgress = true
+	defer func() { m.syncingInProgress = false }()
+
+	epochBlocks := make(map[epoch.Index][]*blockReceived)
+	tangleRoots := make(map[epoch.Index]*smt.SparseMerkleTree)
+	epochSyncEnds := make(map[epoch.Index]*epochSyncEnd)
+
+	// We do not request the start nor the ending epoch, as we know the beginning and the end of the chain.
+	for ei := endEpoch - 1; ei > startEpoch; ei-- {
+		db, _ := database.NewMemDB()
+		tangleRoots[ei] = smt.NewSparseMerkleTree(db.NewStore(), db.NewStore(), lo.PanicOnErr(blake2b.New256(nil)))
+		m.requestEpochBlocks(ei, epoch.ComputeEC(ecChain[ei]))
+	}
+
+	// Counters to check termination conditions.
+	epochsToReceive := endEpoch - startEpoch - 1
+	epochBlocksLeft := make(map[epoch.Index]int64)
+
+	m.log.Debug(">> START")
+
+startReadLoop:
+	for {
+		select {
+		case epochSyncStart := <-m.epochSyncStartChan:
+			ei := epochSyncStart.ei
+			ec := epochSyncStart.ec
+			if ei <= startEpoch || ei > endEpoch-1 {
+				m.log.Warnf("received epoch starter for epoch %d while we are syncing from %d to %d", ei, startEpoch, endEpoch)
+				continue
+			}
+			if ec != epoch.ComputeEC(ecChain[ei]) {
+				m.log.Warnf("received epoch starter on wrong EC chain")
+				continue
+			}
+			if _, exists := epochBlocksLeft[ei]; exists {
+				m.log.Warnf("received starter for already-started epoch %d", ei)
+				continue
+			}
+
+			epochBlocksLeft[ei] = epochSyncStart.blocksCount
+			m.log.Debugw("read epoch block count", "EI", ei, "blockCount", epochBlocksLeft[ei])
+
+			epochsToReceive--
+			m.log.Debugf("epochs left %d", epochsToReceive)
+			if epochsToReceive == 0 {
+				break startReadLoop
+			}
```

Comment:
Seems a bit odd to count the epochs instead of matching them explicitly. But even more so, we won't start processing anything unless we have received all `EpochStarter` packages. Imo it should try to process epochs independently of each other and operate more in a sliding window approach